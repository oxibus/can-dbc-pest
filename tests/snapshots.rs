#![cfg(feature = "encoding")]

use std::borrow::Cow;
use std::fs;
use std::path::Path;

use can_dbc_pest::{decode_cp1252, DbcParser, Rule};
use insta::{assert_debug_snapshot, with_settings};
use pest::Parser;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [
        (
            "tests/shared-test-files/dbc-cantools",
            "dbc-cantools",
            true,
            true,
        ),
        ("tests/shared-test-files/canpy", "canpy", true, true),
        ("tests/opendbc/opendbc/dbc", "opendbc", false, false),
    ];
    for (path, file_name, snapshot, cp1251) in test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path, snapshot, cp1251);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(path, snapshot, cp1251);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let path = path.as_ref();
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    match DbcParser::parse(Rule::file, &buffer) {
        Ok(pairs) => {
            if std::env::var("SKIP_INSTA").is_err()
                && (snapshot || std::env::var("FORCE_INSTA").is_ok())
            {
                assert_debug_snapshot!(file_name, pairs);
            }
        }
        Err(e) => panic!("Failed to parse {file_name}.dbc: {e:#?}"),
    }
}
