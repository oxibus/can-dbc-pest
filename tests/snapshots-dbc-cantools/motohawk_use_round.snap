---
source: tests/snapshots.rs
---
[
    Pair {
        rule: file,
        span: Span {
            str: "VERSION \"1.0\"\n\n\nNS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\nBS_:\n\nBU_: PCM1 FOO\n\n\nBO_ 496 ExampleMessage: 8 PCM1\n SG_ Temperature : 0|12@0- (0.01,250) [229.52|270.47] \"degK\"  PCM1,FOO\n SG_ AverageRadius : 6|6@0+ (0.1,0) [0|5] \"m\" Vector__XXX\n SG_ Enable : 7|1@0+ (1,0) [0|0] \"-\" Vector__XXX\n\n\n\n\nCM_ BO_ 496 \"Example message used as template in MotoHawk models.\";\nBA_DEF_ BO_  \"GenMsgCycleTime\" INT 0 65535;\nBA_DEF_DEF_  \"GenMsgCycleTime\" 0;\n\nVAL_ 496 Enable 0 \"Disabled\" 1 \"Enabled\" ;\n\n\n\n",
            start: 0,
            end: 768,
        },
        inner: [
            Pair {
                rule: version,
                span: Span {
                    str: "VERSION \"1.0\"\n",
                    start: 0,
                    end: 14,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"1.0\"",
                            start: 8,
                            end: 13,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "1.0",
                                    start: 9,
                                    end: 12,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: new_symbols,
                span: Span {
                    str: "NS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\n",
                    start: 16,
                    end: 340,
                },
                inner: [
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "NS_DESC_",
                            start: 24,
                            end: 32,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CM_",
                            start: 34,
                            end: 37,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_",
                            start: 39,
                            end: 46,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_",
                            start: 48,
                            end: 51,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_",
                            start: 53,
                            end: 57,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_DEF_",
                            start: 59,
                            end: 67,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_",
                            start: 69,
                            end: 73,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "FILTER",
                            start: 75,
                            end: 81,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_",
                            start: 83,
                            end: 94,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "EV_DATA_",
                            start: 96,
                            end: 104,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "ENVVAR_DATA_",
                            start: 106,
                            end: 118,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_",
                            start: 120,
                            end: 127,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_VAL_",
                            start: 129,
                            end: 140,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_SGTYPE_",
                            start: 142,
                            end: 156,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_SGTYPE_",
                            start: 158,
                            end: 168,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_TYPE_REF_",
                            start: 170,
                            end: 183,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_TABLE_",
                            start: 185,
                            end: 195,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_GROUP_",
                            start: 197,
                            end: 207,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_VALTYPE_",
                            start: 209,
                            end: 221,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIGTYPE_VALTYPE_",
                            start: 223,
                            end: 239,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BO_TX_BU_",
                            start: 241,
                            end: 250,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_REL_",
                            start: 252,
                            end: 263,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_REL_",
                            start: 265,
                            end: 272,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_REL_",
                            start: 274,
                            end: 289,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_SG_REL_",
                            start: 291,
                            end: 301,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_EV_REL_",
                            start: 303,
                            end: 313,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_BO_REL_",
                            start: 315,
                            end: 325,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SG_MUL_VAL_",
                            start: 327,
                            end: 338,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: bit_timing,
                span: Span {
                    str: "BS_:\n",
                    start: 340,
                    end: 345,
                },
                inner: [],
            },
            Pair {
                rule: nodes,
                span: Span {
                    str: "BU_: PCM1 FOO\n\n\n",
                    start: 346,
                    end: 362,
                },
                inner: [
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "PCM1",
                            start: 351,
                            end: 355,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "FOO",
                            start: 356,
                            end: 359,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: message,
                span: Span {
                    str: "BO_ 496 ExampleMessage: 8 PCM1",
                    start: 362,
                    end: 392,
                },
                inner: [
                    Pair {
                        rule: can_id,
                        span: Span {
                            str: "496",
                            start: 366,
                            end: 369,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident_ext,
                        span: Span {
                            str: "ExampleMessage",
                            start: 370,
                            end: 384,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "8",
                            start: 386,
                            end: 387,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "PCM1",
                            start: 388,
                            end: 392,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal,
                span: Span {
                    str: "SG_ Temperature : 0|12@0- (0.01,250) [229.52|270.47] \"degK\"  PCM1,FOO",
                    start: 394,
                    end: 463,
                },
                inner: [
                    Pair {
                        rule: ident_ext,
                        span: Span {
                            str: "Temperature",
                            start: 398,
                            end: 409,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "0",
                            start: 412,
                            end: 413,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "12",
                            start: 414,
                            end: 416,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "0",
                            start: 417,
                            end: 418,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0.01",
                            start: 421,
                            end: 425,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "250",
                            start: 426,
                            end: 429,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "229.52",
                            start: 432,
                            end: 438,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "270.47",
                            start: 439,
                            end: 445,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"degK\"",
                            start: 447,
                            end: 453,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "degK",
                                    start: 448,
                                    end: 452,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "PCM1",
                            start: 455,
                            end: 459,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "FOO",
                            start: 460,
                            end: 463,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal,
                span: Span {
                    str: "SG_ AverageRadius : 6|6@0+ (0.1,0) [0|5] \"m\" Vector__XXX",
                    start: 465,
                    end: 521,
                },
                inner: [
                    Pair {
                        rule: ident_ext,
                        span: Span {
                            str: "AverageRadius",
                            start: 469,
                            end: 482,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "6",
                            start: 485,
                            end: 486,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "6",
                            start: 487,
                            end: 488,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "0",
                            start: 489,
                            end: 490,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0.1",
                            start: 493,
                            end: 496,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 497,
                            end: 498,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 501,
                            end: 502,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "5",
                            start: 503,
                            end: 504,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"m\"",
                            start: 506,
                            end: 509,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "m",
                                    start: 507,
                                    end: 508,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "Vector__XXX",
                            start: 510,
                            end: 521,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal,
                span: Span {
                    str: "SG_ Enable : 7|1@0+ (1,0) [0|0] \"-\" Vector__XXX",
                    start: 523,
                    end: 570,
                },
                inner: [
                    Pair {
                        rule: ident_ext,
                        span: Span {
                            str: "Enable",
                            start: 527,
                            end: 533,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "7",
                            start: 536,
                            end: 537,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "1",
                            start: 538,
                            end: 539,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "0",
                            start: 540,
                            end: 541,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "1",
                            start: 544,
                            end: 545,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 546,
                            end: 547,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 550,
                            end: 551,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 552,
                            end: 553,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"-\"",
                            start: 555,
                            end: 558,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "-",
                                    start: 556,
                                    end: 557,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "Vector__XXX",
                            start: 559,
                            end: 570,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: comment,
                span: Span {
                    str: "CM_ BO_ 496 \"Example message used as template in MotoHawk models.\";",
                    start: 575,
                    end: 642,
                },
                inner: [
                    Pair {
                        rule: can_id,
                        span: Span {
                            str: "496",
                            start: 583,
                            end: 586,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"Example message used as template in MotoHawk models.\"",
                            start: 587,
                            end: 641,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "Example message used as template in MotoHawk models.",
                                    start: 588,
                                    end: 640,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: attr_def,
                span: Span {
                    str: "BA_DEF_ BO_  \"GenMsgCycleTime\" INT 0 65535;",
                    start: 643,
                    end: 686,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"GenMsgCycleTime\"",
                            start: 656,
                            end: 673,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "GenMsgCycleTime",
                                    start: 657,
                                    end: 672,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 678,
                            end: 679,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "65535",
                            start: 680,
                            end: 685,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: ba_def_def,
                span: Span {
                    str: "BA_DEF_DEF_  \"GenMsgCycleTime\" 0;",
                    start: 687,
                    end: 720,
                },
                inner: [
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"GenMsgCycleTime\"",
                            start: 700,
                            end: 717,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "GenMsgCycleTime",
                                    start: 701,
                                    end: 716,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: number,
                        span: Span {
                            str: "0",
                            start: 718,
                            end: 719,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: value_table_def,
                span: Span {
                    str: "VAL_ 496 Enable 0 \"Disabled\" 1 \"Enabled\" ;",
                    start: 722,
                    end: 764,
                },
                inner: [
                    Pair {
                        rule: can_id,
                        span: Span {
                            str: "496",
                            start: 727,
                            end: 730,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "Enable",
                            start: 731,
                            end: 737,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "0",
                            start: 738,
                            end: 739,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"Disabled\"",
                            start: 740,
                            end: 750,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "Disabled",
                                    start: 741,
                                    end: 749,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "1",
                            start: 751,
                            end: 752,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: string,
                        span: Span {
                            str: "\"Enabled\"",
                            start: 753,
                            end: 762,
                        },
                        inner: [
                            Pair {
                                rule: str,
                                span: Span {
                                    str: "Enabled",
                                    start: 754,
                                    end: 761,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: EOI,
                span: Span {
                    str: "",
                    start: 768,
                    end: 768,
                },
                inner: [],
            },
        ],
    },
]
